// interacto-angular/esm2020/lib/material/mat-interaction-base.mjs
import { InteractionBase } from 'interacto';
import { MatSelect } from '@angular/material/select';
import { MatButtonToggle } from '@angular/material/button-toggle';
import { MatEvent } from './mat-event';
export class MatInteractionBase extends InteractionBase {
    constructor(logger, fsm, data) {
        super(fsm, data, logger);
    }
    registerEventToMatObject(node) {
        if (node instanceof MatSelect) {
            if (this.currentSubscription === undefined) {
                this.currentSubscription = node.selectionChange.subscribe(value => {
                    this.processEvent(new MatEvent(value));
                });
            }
        }
    }
    unregisterEventToMatObject() {
        this.currentSubscription?.unsubscribe();
    }
    onNewNodeRegistered(node) {
        this.getEventTypesOf(this._fsm.currentState).forEach(type => {
            if (node instanceof MatSelect || node instanceof MatButtonToggle) {
                this.registerEventToMatObject(node);
            }
            if (node instanceof EventTarget) {
                this.registerEventToNode(type, node);
                return;
            }
        });
    }
    onNodeUnregistered(node) {
        this.getEventTypesOf(this._fsm.currentState).forEach(type => {
            if (node instanceof EventTarget) {
                this.unregisterEventToNode(type, node);
                return;
            }
            else {
                this.unregisterEventToMatObject();
            }
        });
    }
    reinit() {
        this.unregisterEventToMatObject();
        super.reinit();
    }
    ;
}
