// interacto-angular/esm2020/lib/testing-interacto-angular.module.mjs
import { undoHistoryFactory } from './interacto-angular.module';
import { Bindings, BindingsContext, UndoHistory, UndoHistoryImpl } from 'interacto';
import { NgModule } from '@angular/core';
import { AngularBindings } from './base/angular-bindings';
import * as i0 from "@angular/core";
export function bindingsContextFactory() {
    return new BindingsContext();
}
export function testingBindingsFactory(ctx) {
    const binding = new AngularBindings(new UndoHistoryImpl());
    binding.setBindingObserver(ctx);
    return binding;
}
export class TestingInteractoModule {
}
TestingInteractoModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: TestingInteractoModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TestingInteractoModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.4", ngImport: i0, type: TestingInteractoModule });
TestingInteractoModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: TestingInteractoModule, providers: [{
            provide: BindingsContext,
            useFactory: bindingsContextFactory
        }, {
            provide: AngularBindings,
            useFactory: testingBindingsFactory,
            deps: [
                BindingsContext
            ]
        }, {
            provide: Bindings,
            useExisting: AngularBindings,
        }, {
            provide: UndoHistory,
            useFactory: undoHistoryFactory,
            deps: [
                AngularBindings
            ]
        }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: TestingInteractoModule, decorators: [{
            type: NgModule,
            args: [{
                    providers: [{
                            provide: BindingsContext,
                            useFactory: bindingsContextFactory
                        }, {
                            provide: AngularBindings,
                            useFactory: testingBindingsFactory,
                            deps: [
                                BindingsContext
                            ]
                        }, {
                            provide: Bindings,
                            useExisting: AngularBindings,
                        }, {
                            provide: UndoHistory,
                            useFactory: undoHistoryFactory,
                            deps: [
                                AngularBindings
                            ]
                        }],
                }]
        }] });
