// interacto-angular/esm2020/lib/directives/button-binder.directive.mjs
import { Directive, EventEmitter, Host, Input, Optional, Output } from '@angular/core';
import { InteractoBinderDirective } from './interacto-binder-directive';
import * as i0 from "@angular/core";
import * as i1 from "./on-dynamic.directive";
import * as i2 from "interacto";
export class ButtonBinderDirective extends InteractoBinderDirective {
    constructor(onDyn, element, viewContainerRef, bindings) {
        super(onDyn, element, viewContainerRef);
        this.bindings = bindings;
        this.buttonBinder = new EventEmitter();
    }
    set ioButton(fn) {
        this.callBinder(fn);
    }
    createPartialBinder() {
        return this.bindings.buttonBinder();
    }
    getOutputEvent() {
        return this.buttonBinder;
    }
}
ButtonBinderDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: ButtonBinderDirective, deps: [{ token: i1.OnDynamicDirective, host: true, optional: true }, { token: i0.ElementRef }, { token: i0.ViewContainerRef }, { token: i2.Bindings }], target: i0.ɵɵFactoryTarget.Directive });
ButtonBinderDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.0.4", type: ButtonBinderDirective, selector: "button:[ioButton],[ioButton] [ioOnDynamic]", inputs: { ioButton: "ioButton" }, outputs: { buttonBinder: "buttonBinder" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: ButtonBinderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'button:[ioButton],[ioButton] [ioOnDynamic]'
                }]
        }], ctorParameters: function () { return [{ type: i1.OnDynamicDirective, decorators: [{
                    type: Optional
                }, {
                    type: Host
                }] }, { type: i0.ElementRef }, { type: i0.ViewContainerRef }, { type: i2.Bindings }]; }, propDecorators: { buttonBinder: [{
                type: Output
            }], ioButton: [{
                type: Input
            }] } });
