// interacto-angular/esm2020/lib/directives/clicks-binder.directive.mjs
import { Directive, EventEmitter, Host, Input, Optional, Output } from '@angular/core';
import { InteractoBinderDirective } from './interacto-binder-directive';
import * as i0 from "@angular/core";
import * as i1 from "./on-dynamic.directive";
import * as i2 from "interacto";
export class ClicksBinderDirective extends InteractoBinderDirective {
    constructor(onDyn, element, viewContainerRef, changeDetectorRef, bindings) {
        super(onDyn, element, viewContainerRef, changeDetectorRef);
        this.bindings = bindings;
        /**
         * The expected number of clicks.
         */
        this.count = 2;
        this.clicksBinder = new EventEmitter();
    }
    /**
     * Starts the creation of a binding using the clicks interaction.
     * @param fn - The function of the component that will be called to configure the binding.
     */
    set ioClicks(fn) {
        this.callBinder(fn);
    }
    createPartialBinder() {
        let countValue = typeof this.count === 'number' ? this.count : parseInt(this.count, 0);
        if (isNaN(countValue)) {
            console.log('Cannot create a clicks binder since the value of number is ' +
                'not a number (in string or number format). So using the default value: 2');
            countValue = 2;
        }
        return this.bindings.clicksBinder(countValue);
    }
    getOutputEvent() {
        return this.clicksBinder;
    }
}
ClicksBinderDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: ClicksBinderDirective, deps: [{ token: i1.OnDynamicDirective, host: true, optional: true }, { token: i0.ElementRef }, { token: i0.ViewContainerRef }, { token: i0.ChangeDetectorRef }, { token: i2.Bindings }], target: i0.ɵɵFactoryTarget.Directive });
ClicksBinderDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.0.4", type: ClicksBinderDirective, selector: "[ioClicks]", inputs: { count: "count", ioClicks: "ioClicks" }, outputs: { clicksBinder: "clicksBinder" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: ClicksBinderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ioClicks]'
                }]
        }], ctorParameters: function () { return [{ type: i1.OnDynamicDirective, decorators: [{
                    type: Optional
                }, {
                    type: Host
                }] }, { type: i0.ElementRef }, { type: i0.ViewContainerRef }, { type: i0.ChangeDetectorRef }, { type: i2.Bindings }]; }, propDecorators: { clicksBinder: [{
                type: Output
            }], count: [{
                type: Input
            }], ioClicks: [{
                type: Input
            }] } });
