// interacto-angular/esm2020/lib/directives/drag-lock-binder.directive.mjs
import { Directive, EventEmitter, Host, Input, Optional, Output } from '@angular/core';
import { InteractoBinderDirective } from './interacto-binder-directive';
import * as i0 from "@angular/core";
import * as i1 from "./on-dynamic.directive";
import * as i2 from "interacto";
export class DragLockBinderDirective extends InteractoBinderDirective {
    constructor(onDyn, element, viewContainerRef, bindings) {
        super(onDyn, element, viewContainerRef);
        this.bindings = bindings;
        this.dragLockBinder = new EventEmitter();
    }
    /**
     * Starts the creation of a binding using the drag lock interaction.
     * @param fn - The function of the component that will be called to configure the binding.
     */
    set ioDragLock(fn) {
        this.callBinder(fn);
    }
    createPartialBinder() {
        return this.bindings.dragLockBinder();
    }
    getOutputEvent() {
        return this.dragLockBinder;
    }
}
DragLockBinderDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: DragLockBinderDirective, deps: [{ token: i1.OnDynamicDirective, host: true, optional: true }, { token: i0.ElementRef }, { token: i0.ViewContainerRef }, { token: i2.Bindings }], target: i0.ɵɵFactoryTarget.Directive });
DragLockBinderDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.0.4", type: DragLockBinderDirective, selector: "[ioDragLock]", inputs: { ioDragLock: "ioDragLock" }, outputs: { dragLockBinder: "dragLockBinder" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: DragLockBinderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ioDragLock]'
                }]
        }], ctorParameters: function () { return [{ type: i1.OnDynamicDirective, decorators: [{
                    type: Optional
                }, {
                    type: Host
                }] }, { type: i0.ElementRef }, { type: i0.ViewContainerRef }, { type: i2.Bindings }]; }, propDecorators: { dragLockBinder: [{
                type: Output
            }], ioDragLock: [{
                type: Input
            }] } });
