// interacto-angular/esm2020/lib/directives/focus-on-mouse-enter.directive.mjs
import { Directive } from '@angular/core';
import { FocusHTMLElement } from 'interacto';
import * as i0 from "@angular/core";
import * as i1 from "interacto";
export class FocusOnMouseEnterDirective {
    constructor(element, bindings) {
        this.element = element;
        this.bindings = bindings;
    }
    ngAfterContentInit() {
        this.binding = this.bindings
            .mouseEnterBinder(true)
            .on(this.element)
            .toProduce(i => new FocusHTMLElement(i.currentTarget))
            .bind();
    }
    ngOnDestroy() {
        if (this.binding !== undefined) {
            this.binding.uninstallBinding();
        }
    }
}
FocusOnMouseEnterDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: FocusOnMouseEnterDirective, deps: [{ token: i0.ElementRef }, { token: i1.Bindings }], target: i0.ɵɵFactoryTarget.Directive });
FocusOnMouseEnterDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.0.4", type: FocusOnMouseEnterDirective, selector: "[ioFocusOnMouseEnter]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: FocusOnMouseEnterDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ioFocusOnMouseEnter]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i1.Bindings }]; } });
