// interacto-angular/esm2020/lib/directives/input-binder.directive.mjs
import { Directive, EventEmitter, Host, Input, Optional, Output } from '@angular/core';
import { InteractoBinderDirective } from './interacto-binder-directive';
import * as i0 from "@angular/core";
import * as i1 from "./on-dynamic.directive";
import * as i2 from "interacto";
export class InputBinderDirective extends InteractoBinderDirective {
    constructor(onDyn, element, viewContainerRef, bindings) {
        super(onDyn, element, viewContainerRef);
        this.bindings = bindings;
        this.inputBinder = new EventEmitter();
    }
    set ioInput(fn) {
        this.callBinder(fn);
    }
    createPartialBinder() {
        const elt = this.getElementContent();
        if (elt instanceof HTMLInputElement) {
            switch (elt.type) {
                case 'checkbox':
                case 'radio':
                    return this.bindings.checkboxBinder();
                case 'color':
                    return this.bindings.colorPickerBinder();
                case 'date':
                    return this.bindings.dateBinder();
                case 'number':
                    return this.bindings.spinnerBinder();
            }
        }
        throw new Error('Cannot create a binder on the input.');
    }
    getOutputEvent() {
        return this.inputBinder;
    }
}
InputBinderDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: InputBinderDirective, deps: [{ token: i1.OnDynamicDirective, host: true, optional: true }, { token: i0.ElementRef }, { token: i0.ViewContainerRef }, { token: i2.Bindings }], target: i0.ɵɵFactoryTarget.Directive });
InputBinderDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.0.4", type: InputBinderDirective, selector: "input:[ioInput] :not([type=text]), [ioInput] [ioOnDynamic]", inputs: { ioInput: "ioInput" }, outputs: { inputBinder: "inputBinder" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: InputBinderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'input:[ioInput] :not([type=text]), [ioInput] [ioOnDynamic]'
                }]
        }], ctorParameters: function () { return [{ type: i1.OnDynamicDirective, decorators: [{
                    type: Optional
                }, {
                    type: Host
                }] }, { type: i0.ElementRef }, { type: i0.ViewContainerRef }, { type: i2.Bindings }]; }, propDecorators: { inputBinder: [{
                type: Output
            }], ioInput: [{
                type: Input
            }] } });
