// interacto-angular/esm2020/lib/directives/interacto-binder-directive.mjs
import { Directive, ElementRef } from '@angular/core';
import { BindingImpl } from 'interacto';
import * as i0 from "@angular/core";
/**
 * Base class for Interacto's interactions Directives
 * @typeParam T - The type of the HTML element on which the directive operate
 * @typeParam B - The type of the partial binder the directive will produce
 */
export class InteractoBinderDirective {
    constructor(onDyn, element, viewContainerRef, changeDetectorRef) {
        this.onDyn = onDyn;
        this.element = element;
        this.viewContainerRef = viewContainerRef;
        this.changeDetectorRef = changeDetectorRef;
        this.inputSet = false;
    }
    // tslint:disable-next-line:ban-types
    checkFnName(fn) {
        if (fn === undefined || typeof fn === "string") {
            return undefined;
        }
        return fn?.name;
    }
    callBinder(fn) {
        const fnName = this.checkFnName(fn);
        if (fnName === undefined) {
            return;
        }
        this.inputSet = true;
        // Detects changes to the component and retrieves the input values
        this.changeDetectorRef?.detectChanges();
        const binding = this.getComponent(fnName)[fnName](this.completePartialBinder(), this.getElementContent());
        if (binding instanceof BindingImpl) {
            this.binding = [binding];
        }
        else {
            if (Array.isArray(binding)) {
                this.binding = binding.filter(b => b instanceof BindingImpl).map(b => b);
            }
        }
    }
    getElementContent() {
        return this.element instanceof ElementRef ? this.element.nativeElement : this.element;
    }
    /**
     * Finds the host component. Hack...
     * @param fnName The name of the function to call for building the binding
     */
    getComponent(fnName) {
        // Finding the component. Warning: #horriblehack
        // https://github.com/angular/angular/issues/8277
        // Do not know why '8' (found by inspecting the object at run time)
        return this.viewContainerRef._hostLView[8][fnName] === undefined ?
            // When the directive is used on a template (eg ng For), have to go deeper in the object
            this.viewContainerRef._hostLView[16][8] :
            this.viewContainerRef._hostLView[8];
    }
    ngAfterContentInit() {
        if (!this.inputSet) {
            this.getOutputEvent()?.emit(this.completePartialBinder());
        }
    }
    ngOnDestroy() {
        this.binding?.forEach(b => b.uninstallBinding());
        this.getOutputEvent()?.complete();
    }
    completePartialBinder() {
        const elt = this.getElementContent();
        return (this.onDyn && elt instanceof Node ?
            this.createPartialBinder().onDynamic(elt) :
            this.createPartialBinder().on(this.element));
    }
}
InteractoBinderDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: InteractoBinderDirective, deps: "invalid", target: i0.ɵɵFactoryTarget.Directive });
InteractoBinderDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.0.4", type: InteractoBinderDirective, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: InteractoBinderDirective, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: undefined }, { type: undefined }, { type: i0.ViewContainerRef }, { type: i0.ChangeDetectorRef }]; } });
