// interacto-angular/esm2020/lib/directives/keys-type-binder.directive.mjs
import { Directive, EventEmitter, Host, Input, Optional, Output } from '@angular/core';
import { InteractoBinderDirective } from './interacto-binder-directive';
import * as i0 from "@angular/core";
import * as i1 from "./on-dynamic.directive";
import * as i2 from "interacto";
export class KeysTypeBinderDirective extends InteractoBinderDirective {
    constructor(onDyn, element, viewContainerRef, bindings) {
        super(onDyn, element, viewContainerRef);
        this.bindings = bindings;
        this.keysTypeBinder = new EventEmitter();
    }
    /**
     * Starts the creation of a binding using the multiple key typing interaction.
     * @param fn - The function of the component that will be called to configure the binding.
     */
    set ioKeysType(fn) {
        this.callBinder(fn);
    }
    createPartialBinder() {
        return this.bindings.keysTypeBinder();
    }
    getOutputEvent() {
        return this.keysTypeBinder;
    }
}
KeysTypeBinderDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: KeysTypeBinderDirective, deps: [{ token: i1.OnDynamicDirective, host: true, optional: true }, { token: i0.ElementRef }, { token: i0.ViewContainerRef }, { token: i2.Bindings }], target: i0.ɵɵFactoryTarget.Directive });
KeysTypeBinderDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.0.4", type: KeysTypeBinderDirective, selector: "[ioKeysType]", inputs: { ioKeysType: "ioKeysType" }, outputs: { keysTypeBinder: "keysTypeBinder" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: KeysTypeBinderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ioKeysType]'
                }]
        }], ctorParameters: function () { return [{ type: i1.OnDynamicDirective, decorators: [{
                    type: Optional
                }, {
                    type: Host
                }] }, { type: i0.ElementRef }, { type: i0.ViewContainerRef }, { type: i2.Bindings }]; }, propDecorators: { keysTypeBinder: [{
                type: Output
            }], ioKeysType: [{
                type: Input
            }] } });
