// interacto-angular/esm2020/lib/directives/long-mousedown-binder.directive.mjs
import { Directive, EventEmitter, Host, Input, Optional, Output } from '@angular/core';
import { InteractoBinderDirective } from './interacto-binder-directive';
import * as i0 from "@angular/core";
import * as i1 from "./on-dynamic.directive";
import * as i2 from "interacto";
export class LongMousedownBinderDirective extends InteractoBinderDirective {
    constructor(onDyn, element, viewContainerRef, changeDetectorRef, bindings) {
        super(onDyn, element, viewContainerRef, changeDetectorRef);
        this.bindings = bindings;
        /**
         * The duration of the pressure to end the user interaction.
         * If this duration is not reached, the interaction is cancelled.
         */
        this.duration = 1000;
        this.longMousedownBinder = new EventEmitter();
    }
    /**
     * Starts the creation of a binding using the long press interaction.
     * @param fn - The function of the component that will be called to configure the binding.
     */
    set ioLongMousedown(fn) {
        this.callBinder(fn);
    }
    createPartialBinder() {
        return this.bindings.longMouseDownBinder(this.duration);
    }
    getOutputEvent() {
        return this.longMousedownBinder;
    }
}
LongMousedownBinderDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: LongMousedownBinderDirective, deps: [{ token: i1.OnDynamicDirective, host: true, optional: true }, { token: i0.ElementRef }, { token: i0.ViewContainerRef }, { token: i0.ChangeDetectorRef }, { token: i2.Bindings }], target: i0.ɵɵFactoryTarget.Directive });
LongMousedownBinderDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.0.4", type: LongMousedownBinderDirective, selector: "[ioLongMousedown]", inputs: { duration: "duration", ioLongMousedown: "ioLongMousedown" }, outputs: { longMousedownBinder: "longMousedownBinder" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: LongMousedownBinderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ioLongMousedown]'
                }]
        }], ctorParameters: function () { return [{ type: i1.OnDynamicDirective, decorators: [{
                    type: Optional
                }, {
                    type: Host
                }] }, { type: i0.ElementRef }, { type: i0.ViewContainerRef }, { type: i0.ChangeDetectorRef }, { type: i2.Bindings }]; }, propDecorators: { longMousedownBinder: [{
                type: Output
            }], duration: [{
                type: Input
            }], ioLongMousedown: [{
                type: Input
            }] } });
