// interacto-angular/esm2020/lib/directives/mat-select-binder.directive.mjs
import { Directive, EventEmitter, Host, Input, Optional, Output } from '@angular/core';
import { InteractoBinderDirective } from './interacto-binder-directive';
import * as i0 from "@angular/core";
import * as i1 from "./on-dynamic.directive";
import * as i2 from "../base/angular-bindings";
import * as i3 from "@angular/material/select";
export class MatSelectBinderDirective extends InteractoBinderDirective {
    constructor(onDyn, 
    // element: ElementRef<HTMLElement>,
    viewContainerRef, bindings, 
    // The MatSelect component
    select) {
        super(onDyn, select, viewContainerRef);
        this.bindings = bindings;
        this.matSelectBinder = new EventEmitter();
    }
    set ioMatSelect(fn) {
        this.callBinder(fn);
    }
    createPartialBinder() {
        return this.bindings.matSelectBinder();
    }
    getOutputEvent() {
        return this.matSelectBinder;
    }
}
MatSelectBinderDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: MatSelectBinderDirective, deps: [{ token: i1.OnDynamicDirective, host: true, optional: true }, { token: i0.ViewContainerRef }, { token: i2.AngularBindings }, { token: i3.MatSelect }], target: i0.ɵɵFactoryTarget.Directive });
MatSelectBinderDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.0.4", type: MatSelectBinderDirective, selector: "mat-select:[ioMatSelect],[ioMatSelect] [ioOnDynamic]", inputs: { ioMatSelect: "ioMatSelect" }, outputs: { matSelectBinder: "matSelectBinder" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: MatSelectBinderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'mat-select:[ioMatSelect],[ioMatSelect] [ioOnDynamic]'
                }]
        }], ctorParameters: function () { return [{ type: i1.OnDynamicDirective, decorators: [{
                    type: Optional
                }, {
                    type: Host
                }] }, { type: i0.ViewContainerRef }, { type: i2.AngularBindings }, { type: i3.MatSelect }]; }, propDecorators: { matSelectBinder: [{
                type: Output
            }], ioMatSelect: [{
                type: Input
            }] } });
