// interacto-angular/esm2020/lib/directives/mouseleave-binder.directive.mjs
import { Directive, EventEmitter, Host, Input, Optional, Output } from '@angular/core';
import { InteractoBinderDirective } from './interacto-binder-directive';
import * as i0 from "@angular/core";
import * as i1 from "./on-dynamic.directive";
import * as i2 from "interacto";
export class MouseleaveBinderDirective extends InteractoBinderDirective {
    constructor(onDyn, element, viewContainerRef, changeDetectorRef, bindings) {
        super(onDyn, element, viewContainerRef, changeDetectorRef);
        this.bindings = bindings;
        /**
         * The expected number of clicks.
         */
        this.bubbling = false;
        this.mouseleaveBinder = new EventEmitter();
    }
    /**
     * Starts the creation of a binding using the mouse press interaction.
     * @param fn - The function of the component that will be called to configure the binding.
     */
    set ioMouseleave(fn) {
        this.callBinder(fn);
    }
    createPartialBinder() {
        const withbubbling = typeof this.bubbling === 'boolean' ? this.bubbling : this.bubbling === 'true';
        return this.bindings.mouseLeaveBinder(withbubbling);
    }
    getOutputEvent() {
        return this.mouseleaveBinder;
    }
}
MouseleaveBinderDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: MouseleaveBinderDirective, deps: [{ token: i1.OnDynamicDirective, host: true, optional: true }, { token: i0.ElementRef }, { token: i0.ViewContainerRef }, { token: i0.ChangeDetectorRef }, { token: i2.Bindings }], target: i0.ɵɵFactoryTarget.Directive });
MouseleaveBinderDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.0.4", type: MouseleaveBinderDirective, selector: "[ioMouseleave]", inputs: { bubbling: "bubbling", ioMouseleave: "ioMouseleave" }, outputs: { mouseleaveBinder: "mouseleaveBinder" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: MouseleaveBinderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ioMouseleave]'
                }]
        }], ctorParameters: function () { return [{ type: i1.OnDynamicDirective, decorators: [{
                    type: Optional
                }, {
                    type: Host
                }] }, { type: i0.ElementRef }, { type: i0.ViewContainerRef }, { type: i0.ChangeDetectorRef }, { type: i2.Bindings }]; }, propDecorators: { mouseleaveBinder: [{
                type: Output
            }], bubbling: [{
                type: Input
            }], ioMouseleave: [{
                type: Input
            }] } });
