// interacto-angular/esm2020/lib/directives/multi-touch-binder.directive.mjs
import { Directive, EventEmitter, Host, Input, Optional, Output } from '@angular/core';
import { InteractoBinderDirective } from './interacto-binder-directive';
import * as i0 from "@angular/core";
import * as i1 from "./on-dynamic.directive";
import * as i2 from "interacto";
export class MultiTouchBinderDirective extends InteractoBinderDirective {
    constructor(onDyn, element, viewContainerRef, changeDetectorRef, bindings) {
        super(onDyn, element, viewContainerRef, changeDetectorRef);
        this.bindings = bindings;
        /**
         * The number of required touches.
         * A multi-touch starts when all its touches have started.
         * A multi-touch ends when the number of required touches is greater than the number of touches.
         */
        this.nbTouches = 2;
        this.multiTouchBinder = new EventEmitter();
    }
    /**
     * Starts the creation of a binding using the multi touch interaction.
     * @param fn - The function of the component that will be called to configure the binding.
     */
    set ioMultiTouch(fn) {
        this.callBinder(fn);
    }
    createPartialBinder() {
        return this.bindings.multiTouchBinder(this.nbTouches);
    }
    getOutputEvent() {
        return this.multiTouchBinder;
    }
}
MultiTouchBinderDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: MultiTouchBinderDirective, deps: [{ token: i1.OnDynamicDirective, host: true, optional: true }, { token: i0.ElementRef }, { token: i0.ViewContainerRef }, { token: i0.ChangeDetectorRef }, { token: i2.Bindings }], target: i0.ɵɵFactoryTarget.Directive });
MultiTouchBinderDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.0.4", type: MultiTouchBinderDirective, selector: "[ioMultiTouch]", inputs: { nbTouches: "nbTouches", ioMultiTouch: "ioMultiTouch" }, outputs: { multiTouchBinder: "multiTouchBinder" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: MultiTouchBinderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ioMultiTouch]'
                }]
        }], ctorParameters: function () { return [{ type: i1.OnDynamicDirective, decorators: [{
                    type: Optional
                }, {
                    type: Host
                }] }, { type: i0.ElementRef }, { type: i0.ViewContainerRef }, { type: i0.ChangeDetectorRef }, { type: i2.Bindings }]; }, propDecorators: { multiTouchBinder: [{
                type: Output
            }], nbTouches: [{
                type: Input
            }], ioMultiTouch: [{
                type: Input
            }] } });
