// interacto-angular/esm2020/lib/directives/pan-binder.directive.mjs
import { Directive, EventEmitter, Host, Input, Optional, Output } from '@angular/core';
import { InteractoBinderDirective } from './interacto-binder-directive';
import * as i0 from "@angular/core";
import * as i1 from "./on-dynamic.directive";
import * as i2 from "interacto";
export class PanBinderDirective extends InteractoBinderDirective {
    constructor(onDyn, element, viewContainerRef, changeDetectorRef, bindings) {
        super(onDyn, element, viewContainerRef, changeDetectorRef);
        this.bindings = bindings;
        /**
         * Defines whether the pan is horizontal or vertical
         */
        this.horizontal = true;
        /**
         * The minimal distance from the starting point to the release point for validating the pan
         */
        this.minLength = 100;
        /**
         * The tolerance rate in pixels accepted while executing the pan
         */
        this.pxTolerance = 20;
        /**
         * The number of touches required to start the interaction
         */
        this.nbTouches = 1;
        this.panBinder = new EventEmitter();
    }
    /**
     * Starts the creation of a binding using the pan interaction.
     * @param fn - The function of the component that will be called to configure the binding.
     */
    set ioPan(fn) {
        this.callBinder(fn);
    }
    createPartialBinder() {
        return this.bindings.panBinder(this.horizontal, this.minLength, this.nbTouches, this.pxTolerance);
    }
    getOutputEvent() {
        return this.panBinder;
    }
}
PanBinderDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: PanBinderDirective, deps: [{ token: i1.OnDynamicDirective, host: true, optional: true }, { token: i0.ElementRef }, { token: i0.ViewContainerRef }, { token: i0.ChangeDetectorRef }, { token: i2.Bindings }], target: i0.ɵɵFactoryTarget.Directive });
PanBinderDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.0.4", type: PanBinderDirective, selector: "[ioPan]", inputs: { horizontal: "horizontal", minLength: "minLength", pxTolerance: "pxTolerance", nbTouches: "nbTouches", ioPan: "ioPan" }, outputs: { panBinder: "panBinder" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: PanBinderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ioPan]'
                }]
        }], ctorParameters: function () { return [{ type: i1.OnDynamicDirective, decorators: [{
                    type: Optional
                }, {
                    type: Host
                }] }, { type: i0.ElementRef }, { type: i0.ViewContainerRef }, { type: i0.ChangeDetectorRef }, { type: i2.Bindings }]; }, propDecorators: { panBinder: [{
                type: Output
            }], horizontal: [{
                type: Input
            }], minLength: [{
                type: Input
            }], pxTolerance: [{
                type: Input
            }], nbTouches: [{
                type: Input
            }], ioPan: [{
                type: Input
            }] } });
