// interacto-angular/esm2020/lib/directives/redo-binder.directive.mjs
import { Directive, Input } from '@angular/core';
import { Redo } from 'interacto';
import { InteractoBinderDirective } from './interacto-binder-directive';
import * as i0 from "@angular/core";
import * as i1 from "interacto";
export class RedoBinderDirective extends InteractoBinderDirective {
    constructor(element, viewContainerRef, bindings) {
        super(undefined, element, viewContainerRef);
        this.bindings = bindings;
        this.redoObs = bindings.undoHistory
            .redosObservable()
            .subscribe(_ => {
            this.updateRedo();
        });
    }
    set ioRedo(fn) {
        this.callBinder(fn);
    }
    updateRedo() {
        const elt = this.getElementContent();
        elt.disabled = this.bindings.undoHistory.getLastRedo() === undefined;
        elt.title = this.bindings.undoHistory.getLastOrEmptyRedoMessage();
    }
    ngAfterContentInit() {
        this.updateRedo();
        if (!this.inputSet) {
            const b = this.createPartialBinder()
                .catch(err => {
                console.log(err);
            })
                .bind();
            this.binding = [b];
        }
    }
    createPartialBinder() {
        return this.bindings
            .buttonBinder()
            .on(this.element)
            .toProduce(() => new Redo(this.bindings.undoHistory));
    }
    ngOnDestroy() {
        super.ngOnDestroy();
        this.redoObs.unsubscribe();
    }
    getOutputEvent() {
        return undefined;
    }
}
RedoBinderDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: RedoBinderDirective, deps: [{ token: i0.ElementRef }, { token: i0.ViewContainerRef }, { token: i1.Bindings }], target: i0.ɵɵFactoryTarget.Directive });
RedoBinderDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.0.4", type: RedoBinderDirective, selector: "button:ioRedo,button:[ioRedo]", inputs: { ioRedo: "ioRedo" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: RedoBinderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'button:ioRedo,button:[ioRedo]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.ViewContainerRef }, { type: i1.Bindings }]; }, propDecorators: { ioRedo: [{
                type: Input
            }] } });
