// interacto-angular/esm2020/lib/directives/select-binder.directive.mjs
import { Directive, EventEmitter, Host, Input, Optional, Output } from '@angular/core';
import { InteractoBinderDirective } from './interacto-binder-directive';
import * as i0 from "@angular/core";
import * as i1 from "./on-dynamic.directive";
import * as i2 from "interacto";
export class SelectBinderDirective extends InteractoBinderDirective {
    constructor(onDyn, element, viewContainerRef, bindings) {
        super(onDyn, element, viewContainerRef);
        this.bindings = bindings;
        this.selectBinder = new EventEmitter();
    }
    set ioSelect(fn) {
        this.callBinder(fn);
    }
    createPartialBinder() {
        return this.bindings.comboBoxBinder();
    }
    getOutputEvent() {
        return this.selectBinder;
    }
}
SelectBinderDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: SelectBinderDirective, deps: [{ token: i1.OnDynamicDirective, host: true, optional: true }, { token: i0.ElementRef }, { token: i0.ViewContainerRef }, { token: i2.Bindings }], target: i0.ɵɵFactoryTarget.Directive });
SelectBinderDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.0.4", type: SelectBinderDirective, selector: "select:[ioSelect], [ioSelect] [ioOnDynamic]", inputs: { ioSelect: "ioSelect" }, outputs: { selectBinder: "selectBinder" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: SelectBinderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'select:[ioSelect], [ioSelect] [ioOnDynamic]'
                }]
        }], ctorParameters: function () { return [{ type: i1.OnDynamicDirective, decorators: [{
                    type: Optional
                }, {
                    type: Host
                }] }, { type: i0.ElementRef }, { type: i0.ViewContainerRef }, { type: i2.Bindings }]; }, propDecorators: { selectBinder: [{
                type: Output
            }], ioSelect: [{
                type: Input
            }] } });
