// interacto-angular/esm2020/lib/directives/swipe-binder.directive.mjs
import { Directive, EventEmitter, Host, Input, Optional, Output } from '@angular/core';
import { InteractoBinderDirective } from './interacto-binder-directive';
import * as i0 from "@angular/core";
import * as i1 from "./on-dynamic.directive";
import * as i2 from "interacto";
export class SwipeBinderDirective extends InteractoBinderDirective {
    constructor(onDyn, element, viewContainerRef, changeDetectorRef, bindings) {
        super(onDyn, element, viewContainerRef, changeDetectorRef);
        this.bindings = bindings;
        /**
         * Defines whether the swipe is horizontal or vertical
         */
        this.horizontal = true;
        /**
         * The minimal minVelocity to reach for validating the swipe. In pixels per second.
         */
        this.minVelocity = 100;
        /**
         * The minimal distance from the starting point to the release point for validating the swipe
         */
        this.minLength = 100;
        /**
         * The tolerance rate in pixels accepted while executing the swipe
         */
        this.pxTolerance = 20;
        /**
         * The number of touches required to start the interaction
         */
        this.nbTouches = 1;
        this.swipeBinder = new EventEmitter();
    }
    /**
     * Starts the creation of a binding using the swipe interaction.
     * @param fn - The function of the component that will be called to configure the binding.
     */
    set ioSwipe(fn) {
        this.callBinder(fn);
    }
    createPartialBinder() {
        return this.bindings.swipeBinder(this.horizontal, this.minVelocity, this.minLength, this.nbTouches, this.pxTolerance);
    }
    getOutputEvent() {
        return this.swipeBinder;
    }
}
SwipeBinderDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: SwipeBinderDirective, deps: [{ token: i1.OnDynamicDirective, host: true, optional: true }, { token: i0.ElementRef }, { token: i0.ViewContainerRef }, { token: i0.ChangeDetectorRef }, { token: i2.Bindings }], target: i0.ɵɵFactoryTarget.Directive });
SwipeBinderDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.0.4", type: SwipeBinderDirective, selector: "[ioSwipe]", inputs: { horizontal: "horizontal", minVelocity: "minVelocity", minLength: "minLength", pxTolerance: "pxTolerance", nbTouches: "nbTouches", ioSwipe: "ioSwipe" }, outputs: { swipeBinder: "swipeBinder" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: SwipeBinderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ioSwipe]'
                }]
        }], ctorParameters: function () { return [{ type: i1.OnDynamicDirective, decorators: [{
                    type: Optional
                }, {
                    type: Host
                }] }, { type: i0.ElementRef }, { type: i0.ViewContainerRef }, { type: i0.ChangeDetectorRef }, { type: i2.Bindings }]; }, propDecorators: { swipeBinder: [{
                type: Output
            }], horizontal: [{
                type: Input
            }], minVelocity: [{
                type: Input
            }], minLength: [{
                type: Input
            }], pxTolerance: [{
                type: Input
            }], nbTouches: [{
                type: Input
            }], ioSwipe: [{
                type: Input
            }] } });
