// interacto-angular/esm2020/lib/directives/tap-binder.directive.mjs
import { Directive, EventEmitter, Host, Input, Optional, Output } from '@angular/core';
import { InteractoBinderDirective } from './interacto-binder-directive';
import * as i0 from "@angular/core";
import * as i1 from "./on-dynamic.directive";
import * as i2 from "interacto";
export class TapBinderDirective extends InteractoBinderDirective {
    constructor(onDyn, element, viewContainerRef, changeDetectorRef, bindings) {
        super(onDyn, element, viewContainerRef, changeDetectorRef);
        this.bindings = bindings;
        /**
         * The number of taps expected to end the interaction.
         * If this number is not reached after a timeout, the interaction is cancelled.
         */
        this.nbTaps = 2;
        this.tapBinder = new EventEmitter();
    }
    /**
     * Starts the creation of a binding using the tap interaction.
     * @param fn - The function of the component that will be called to configure the binding.
     */
    set ioTap(fn) {
        this.callBinder(fn);
    }
    createPartialBinder() {
        return this.bindings.tapBinder(this.nbTaps);
    }
    getOutputEvent() {
        return this.tapBinder;
    }
}
TapBinderDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: TapBinderDirective, deps: [{ token: i1.OnDynamicDirective, host: true, optional: true }, { token: i0.ElementRef }, { token: i0.ViewContainerRef }, { token: i0.ChangeDetectorRef }, { token: i2.Bindings }], target: i0.ɵɵFactoryTarget.Directive });
TapBinderDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.0.4", type: TapBinderDirective, selector: "[ioTap]", inputs: { nbTaps: "nbTaps", ioTap: "ioTap" }, outputs: { tapBinder: "tapBinder" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: TapBinderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ioTap]'
                }]
        }], ctorParameters: function () { return [{ type: i1.OnDynamicDirective, decorators: [{
                    type: Optional
                }, {
                    type: Host
                }] }, { type: i0.ElementRef }, { type: i0.ViewContainerRef }, { type: i0.ChangeDetectorRef }, { type: i2.Bindings }]; }, propDecorators: { tapBinder: [{
                type: Output
            }], nbTaps: [{
                type: Input
            }], ioTap: [{
                type: Input
            }] } });
