// interacto-angular/esm2020/lib/directives/undo-binder.directive.mjs
import { Directive, Input } from '@angular/core';
import { Undo } from 'interacto';
import { InteractoBinderDirective } from './interacto-binder-directive';
import * as i0 from "@angular/core";
import * as i1 from "interacto";
export class UndoBinderDirective extends InteractoBinderDirective {
    constructor(element, viewContainerRef, bindings) {
        super(undefined, element, viewContainerRef);
        this.bindings = bindings;
        this.undoObs = bindings.undoHistory
            .undosObservable()
            .subscribe(_ => {
            this.updateUndo();
        });
    }
    set ioUndo(fn) {
        this.callBinder(fn);
    }
    updateUndo() {
        const elt = this.getElementContent();
        elt.disabled = this.bindings.undoHistory.getLastUndo() === undefined;
        elt.title = this.bindings.undoHistory.getLastOrEmptyUndoMessage();
    }
    ngAfterContentInit() {
        this.updateUndo();
        if (!this.inputSet) {
            const b = this.createPartialBinder()
                .catch(err => {
                console.log(err);
            })
                .bind();
            this.binding = [b];
        }
    }
    createPartialBinder() {
        return this.bindings
            .buttonBinder()
            .on(this.element)
            .toProduce(() => new Undo(this.bindings.undoHistory));
    }
    ngOnDestroy() {
        super.ngOnDestroy();
        this.undoObs.unsubscribe();
    }
    getOutputEvent() {
        return undefined;
    }
}
UndoBinderDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: UndoBinderDirective, deps: [{ token: i0.ElementRef }, { token: i0.ViewContainerRef }, { token: i1.Bindings }], target: i0.ɵɵFactoryTarget.Directive });
UndoBinderDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.0.4", type: UndoBinderDirective, selector: "button:ioUndo,button:[ioUndo]", inputs: { ioUndo: "ioUndo" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.4", ngImport: i0, type: UndoBinderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'button:ioUndo,button:[ioUndo]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.ViewContainerRef }, { type: i1.Bindings }]; }, propDecorators: { ioUndo: [{
                type: Input
            }] } });
